### PR TITLE
Fix a data race in b2BlockAllocator::b2BlockAllocator()

### PR DESCRIPTION
**Pull request by tlemo**
Fix for #527 

This allows multiple threads to safely use their own (not shared)
b2World instances.

The data race is avoided by taking advantage of the thread-safety
guarantees provided by static local variable initialization in C++11:
https://en.cppreference.com/w/cpp/language/storage_duration#Static_local_variables

### CODE REVIEW

**Comment by jhoffman0x** [comment]
This would fix the data race but it would cause static initialization order problems.

If a user has a static or global object that creates bodies in in its constructor, then the lookup table could be used before its initialized.

**Comment by tlemo** [comment]
@jhoffman0x : that's a good observation, although:

1. The proposed fix doesn't change the initialization order (it still happens inside the b2BlockAllocator's constructor, just in a thread safe manner). So it wouldn't introduce any new issues related to the initialization order.

2. b2BlockAllocator is b2World data member. In order to create any bodies, you need to instantiate a b2World first, which means that the piece relevant to this fix would run. 

To sum up: yes, creating box2d objects from a dynamic initializer is a bad idea in general, but nothing related to the proposed fix. (a iffy part is the existing b2BlockAllocator::s_blockSizes, but except for a compiler flaw even that should be statically initialized)

**Comment by jhoffman0x** [comment]
Ah, yeah I see. I was looking at the diff on github and with the parts cut off around the changes I didn't notice that the static bool was now in the constructor.

**Comment by erincatto** [comment]
Fixed in #583
